src/entt/entity/poly_storage_mixin.hpp

```diff
@@ -0,0 +1,57 @@
+#ifndef ENTT_ENTITY_POLY_STORAGE_MIXIN_HPP
+#define ENTT_ENTITY_POLY_STORAGE_MIXIN_HPP
+
+#include "storage.hpp"
+#include "poly_type_traits.hpp"
+#include "sigh_storage_mixin.hpp"
+
+
+namespace entt {
+
+template<typename Entity, typename Type>
+class poly_type;
+
+/**
+ * @brief Storage mixin for polymorphic component types
+ * @tparam Storage underlying storage type
+ * @tparam Entity entity type
+ * @tparam Type value type
+ */
+template<typename Storage, typename = void>
+class poly_storage_mixin : public Storage {
+public:
+    /*! @brief Underlying value type. */
+    using value_type = typename Storage::value_type;
+    /*! @brief Underlying entity identifier. */
+    using entity_type = typename Storage::entity_type;
+
+    /*! @brief Inherited constructors. */
+    using Storage::Storage;
+
+    /**
+     * @brief Forwards variables to mixins, if any.
+     * @param value A variable wrapped in an opaque container.
+     */
+    void bind(any value) ENTT_NOEXCEPT override {
+        if(auto *reg = any_cast<basic_registry<entity_type>>(&value); reg) {
+            owner = reg;
+            bind_all_parent_types(*reg, poly_parent_types_t<value_type>{});
+        }
+        Storage::bind(std::move(value));
+    }
+
+private:
+    template<typename... ParentTypes>
+    void bind_all_parent_types(basic_registry<entity_type>& reg, [[maybe_unused]] type_list<ParentTypes...>) {
+        (reg.ctx().template emplace<poly_type<entity_type, ParentTypes>>().bind_child_storage(this), ...);
+        reg.ctx().template emplace<poly_type<entity_type, value_type>>().bind_child_storage(this);
```

Comment:
I'm not sure what do you mean here exactly, sorry. Are you asking, why I'm binding child storage for the `value_type` here?

Reply:
Yeah, sorry, my fault. I meant the `value_type` part indeed.

Reply:
It is done because when we are iterating through components, derived from `Type`, the instances of `Type` itself must be among them.

Reply:
Mmm is it? This is another assumption on which I don't have control as an user and that I don't see as mandatory necessarily. 🤔

Reply:
Mmm probably I don't get it correctly. Consider these where I've `circle` that derives from `shape`:
```cpp
(reg.ctx().template emplace<poly_type<entity_type, ParentTypes>>().bind_child_storage(this), ...);
reg.ctx().template emplace<poly_type<entity_type, value_type>>().bind_child_storage(this);
```
The first line registers `circle` with the poly type for shape, right? While the second one registers circle with the poly type if `circle` itself. What would be the use case for it? I still don't get it but I'm likely missing something?

Reply:
For your example let's assume, that there is another class, derived from `circle`. I want to iterate each `circle` and anything that derives from it. To do this, I get `poly_type<entity, circle>` and iterate every pool, that was bound as child to it. For this, to include instances of `circle` itself, the `circle` pool must be bound as a child pool for `circle`.

Reply:
Doesn't it confirm that there are multiple use cases and no choice? 🤔

Reply:
Yeah, sorry, I was out for the weekend and I was in a rush while trying to catch up with all the other things today.

Imagine you have `shape` (the base), `circle` and `rectangle`. What I end up after a while is:

* `poly_type<entity_type, shape>>`
* `poly_type<entity_type, circle>>`
* `poly_type<entity_type, rectangle>>`

What I don't get is why the last two exist and what their purpose is. They don't have derived classes (let's make them `final`) and all I want is to iterate my shapes. Therefore, I can see the reason for the first `poly_type` while the others look useless.
Does it make sense now?

Reply:
Oh, I think I understand now. To be sure: you are concerned, that while some poly types don't have any children and could be treated as a non-polymorphic ones, they still put `poly_type` for themselves into the context, and internally it allocates a vector of at least one element, so we still pay for having them?

Well, at the compile time we cannot be sure, that `shape` and `circle` don't have any children, if they aren't declared final. They may have derived types or they may not, we cannot tell this. Having `poly_type` allows us to iterate every pool, containing this type, it will work fine with or without any derived types. So having `poly_type` for each polymorphic type is very convenient, and, moreover, it is intuitive to have it for all types.

Of course, at runtime, we can add `poly_type` only for types, that have children, and then in each algorithm we can check, if we have a `poly_type` for given type, and, if we don't, fallback to non-polymorphic algorithm. But this solution breaks a really nice abstraction, that does not require polymorphic type data to be stored in the registry, because we will be required to use registry in the fallback case anyway. And the algorithms will become bigger and uglier, all of this for no reasonable gain. 

However, I think I know how to resolve this. Instead of binding storage as a child for its own `poly_type`, which will internally cause heap allocation by the vector, we can store it separately in some, for example, `std::optional<poly_type_holder> self_type_holder`. In this case we still pay for polymorphic type with no children, compared to non-polymorphic type, but this cost is much smaller, than before, and anything else is not broken. What do you think?